test-data/unit/check-basic.test

```diff
@@ -390,3 +390,61 @@ y = x # E: Incompatible types in assignment (expression has type "Dict[str, int]
      # N: "Dict" is invariant -- see http://mypy.readthedocs.io/en/latest/common_issues.html#variance \
      # N: Consider using "Mapping" instead, which is covariant in the value type
 [builtins fixtures/dict.pyi]
+
+[case testDistinctTypes]
+# flags: --strict-optional
+import b
+
+[file a.py]
+from typing import NamedTuple
+class A: pass
+N = NamedTuple('N', [('x', int)])
+
+[file b.py]
+from typing import List, Optional, Union, Sequence, NamedTuple, Tuple, Type
+import a
+class A: pass
+N = NamedTuple('N', [('x', int)])
+
+def foo() -> Optional[A]:
+    b = True
+    return a.A() if b else None  # E: Incompatible return value type (got "Optional[a.A]", expected "Optional[b.A]")
+
+def bar() -> List[A]:
+    l = [a.A()]
+    return l  # E: Incompatible return value type (got "List[a.A]", expected "List[b.A]")
+
+def baz() -> Union[A, int]:
+    b = True
+    return a.A() if b else 10  # E: Incompatible return value type (got "Union[a.A, int]", expected "Union[b.A, int]")
+
+def spam() -> Optional[A]:
+    return a.A()  # E: Incompatible return value type (got "a.A", expected "Optional[b.A]")
+
+def eggs() -> Sequence[A]:
+    x = [a.A()]
+    return x  # E: Incompatible return value type (got "List[a.A]", expected "Sequence[b.A]")
+
+def eggs2() -> Sequence[N]:
+    x = [a.N(0)]
+    return x  # E: Incompatible return value type (got "List[a.N]", expected "Sequence[b.N]")
+
+def asdf1() -> Sequence[Tuple[a.A, A]]:
+    x = [(a.A(), a.A())]
+    return x  # E: Incompatible return value type (got "List[Tuple[a.A, a.A]]", expected "Sequence[Tuple[a.A, b.A]]")
+
+def asdf2() -> Sequence[Tuple[A, a.A]]:
+    x = [(a.A(), a.A())]
+    return x  # E: Incompatible return value type (got "List[Tuple[a.A, a.A]]", expected "Sequence[Tuple[b.A, a.A]]")
+
+def arg() -> Tuple[A, A]:
+    return A()  # E: Incompatible return value type (got "A", expected "Tuple[A, A]")
+
+def types() -> Sequence[Type[A]]:
+    x = [a.A]
+    return x  # E: Incompatible return value type (got "List[Type[a.A]]", expected "Sequence[Type[b.A]]")
+
+a = (a.A(), A())
+a.x  # E: "Tuple[a.A, b.A]" has no attribute "x"
+
+[builtins fixtures/list.pyi]
```

Comment:
I didn't read the whole diff yet, but seeing this set of tests makes me wonder, what if one of the types is `a.A` (i.e. not the locally defined `A`) and the other is, say, `int`? Will it correctly show `a.A`, or might it show just `A`? (Which would be confusing because in scope, `A` refers to `b.A`.)

Reply:
We would just show `A`, yeah. This is a bummer but fixing it would be much more invasive than this diff, since type formatting will need to be much more contextual.